aten/src/ATen/native/sparse/SparseCsrTensor.cpp

```diff
@@ -750,41 +749,55 @@ Tensor select_sparse_csr(const Tensor& self, int64_t dim, int64_t index) {
-          new_sizes,
-          optTypeMetaToScalarType(options.dtype_opt()),
-          options.layout_opt(),
-          options.device_opt(),
-          options.pinned_memory_opt());
-    }
-    return at::native::_sparse_csr_tensor_unsafe(
-        self.crow_indices().select(dim, index),
-        self.col_indices().select(dim, index),
+  Tensor plain_indices;
+  Tensor compressed_indices;
+  std::tie(compressed_indices, plain_indices) =
+      AT_DISPATCH_ROW_SPARSE_COMPRESSED_LAYOUTS(
+          self.layout(),
+          "select",
+          [&]() {
+            return std::make_pair(self.crow_indices(), self.col_indices());
+          },
+          [&]() {
+            return std::make_pair(self.ccol_indices(), self.row_indices());
+          });
+  auto n_batch = compressed_indices.dim() - 1;
+
+  if (dim < n_batch) {
+    // Selecting batch dimension
+    return at::native::_sparse_compressed_tensor_unsafe(
+        compressed_indices.select(dim, index),
+        plain_indices.select(dim, index),
         self.values().select(dim, index),
         new_sizes,
         optTypeMetaToScalarType(options.dtype_opt()),
         options.layout_opt(),
         options.device_opt(),
         options.pinned_memory_opt());
-  } else {
+  } else if (dim < n_batch + 2) {
+    // Selecting sparse dimension
     TORCH_CHECK(
-        self.is_sparse_csr(),
-        "select(): selecting non-batch dimensions is currently only supported for CSR tensors.");
+        self.layout() == kSparseCsr || self.layout() == kSparseCsc,
+        "select(): selecting non-batch dimensions is currently only supported for non-blocked sparse compressed layouts tensors.");
     TORCH_CHECK(
         self.dim() == 2,
-        "select(): selecting rows or columns is not implemented for batched sparse CSR tensors.")
-    // Converting to COO and calling select is slighly slower than operating on
-    // the CSR indices directly for constructing a COO vector, however current
-    // version is more readable and easier to understand.
+        "select(): selecting rows or columns is not implemented for batched sparse compressed tensors.")
+    // Converting to COO and calling select is slightly slower than operating
+    // on the CSR indices directly for constructing a COO vector, however
+    // current version is more readable and easier to understand.
     return self.to_sparse().select(dim, index);
+  } else {
+    // Selecting dense dimension
+    return AT_DISPATCH_PLAIN_SPARSE_COMPRESSED_LAYOUTS(
+        self.layout(),
+        "select",
+        // Non blocked layout (2 sparse dims become 1 nnz dim in values, so dim
+        // is found one position to the left)
+        [&]() { return self.values().select(dim - 1, index); },
+        // Block layout (2 sparse dims become 1 nnz dim + 2 block-shape dims in
+        // values, so dim is found 1 position to the right)
+        [&]() { return self.values().select(dim + 1, index); });
```

Comment:
Shouldn't we subtract batch dims here?

Reply:
If you line up `self.sizes()` and `self.values().sizes()`

```
self.sizes() : (batch[0], ... batch[M-1], | sparse_row, sparse_col, | dense[0],... dense[N-1])
values.sizes(): (batch[0], ... batch[M-1], | nnz, | dense[0], ... dense[N-1]) # csr/csc
values.sizes(): (batch[0], ... batch[M-1], | nnz, block[0], block[1], | dense[0], ... dense[N-1]) # bsr/bsc
```
All dims to the left of `|..|` are preserved in values *and* map directly, all dims to the right of `|..|` are also preserved but get shifted by the fact that the sparse dims are compressed, or compressed and blocked in values.  

In non-blocked layouts the 2 sparse dims in `self` are mapped to the single `nnz` length dim in values, so the input `dim` (in self's shape) is found in values one position back.

In blocked layouts the 2 sparse dims in `self` are mapped to the three `(nnz, ) + blockshape` dims in values, so the input `dim` (in self's) shape is found in values one position forward.

Reply:
@amjames , I think this comment belongs to https://github.com/pytorch/pytorch/pull/82122#discussion_r938086381, right?